netty/src/main/java/io/grpc/netty/NettyServerHandler.java

```diff
@@ -723,6 +722,21 @@ private void cancelStream(ChannelHandlerContext ctx, CancelServerStreamCommand c
     }
   }
 
+  private void gracefulClose(final ChannelHandlerContext ctx, final GracefulServerCloseCommand msg,
+      ChannelPromise promise) throws Exception {
+    // Ideally we'd adjust a pre-existing graceful shutdown's grace period to at least what is
+    // requested here. But that's an edge case and seems bug-prone.
+    if (gracefulShutdown == null) {
+      Long graceTimeInNanos = null;
+      if (msg.getGraceTimeUnit() != null) {
+        graceTimeInNanos = msg.getGraceTimeUnit().toNanos(msg.getGraceTime());
+      }
+      gracefulShutdown = new GracefulShutdown(msg.getGoAwayDebugString(), graceTimeInNanos);
+      gracefulShutdown.start(ctx);
+    }
+    promise.setSuccess();
```

Comment:
Now that `close` is calling `gracefulClose` this changes the behavior. The old `close()` didn't touch `promise`. It could be `null` or calling `promise.setSuccess()` now changes the behavior? Is that intended?

Reply:
The old code not touching the promise was a bug. You must always complete the promise (either successfully or with an error).